In [1]:
from __future__ import unicode_literals

import pandas as pd
import numpy as np
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

# 1. Preparing data - Latest 12 weeks of life of the banana plant
---

In [3]:
wind_speed_l12w = pd.read_csv('../../../../data/raw/FincaPorvenir/Metereologico/Latest_12-weeks_May-4_August-03_2018/' \
                                 'Velocidad-del-viento_May-4_August-03_2018.csv', )

In [4]:
print(wind_speed_l12w.shape)
wind_speed_l12w.head()

(4398, 2)


Fecha:  Velocidad del viento (Km/h)
0  2018-05-04 00:07:38                          0.0
1  2018-05-04 00:37:39                          0.0
2  2018-05-04 01:07:38                          0.0
3  2018-05-04 01:37:38                          0.0
4  2018-05-04 02:07:38                          0.0

## 1.1. We evaluate if this dataset has null type `NaN`
---

In [5]:
print(wind_speed_l12w.isnull().any())
wind_speed_l12w.isnull().values.any()

Fecha:                         False
Velocidad del viento (Km/h)    False
dtype: bool


False

Don't have null values

## 1.2. Selecting  relevant index columns features
---

Since the dataset has a column called **`Fecha:`**, which is not a numerical value,

it will be removed so that it does not interfere **with our subsequent scaling**, 

so we are only going to reference the values or samples of the column

** `Velocidad del viento (Km/h)` ** and assigning them to the matrix 

`wind_speed_l12w_array` created such as follow:

In [6]:
wind_speed_l12w_array = wind_speed_l12w.iloc[:,1].values

In [7]:
# Rehape the luminosity_luxes array 
wind_speed_l12w_array = wind_speed_l12w_array.reshape(-1, 1)
wind_speed_l12w_array

array([[0.        ],
       [0.        ],
       [0.        ],
       ...,
       [0.        ],
       [0.        ],
       [4.21052632]])

## 1.3 Generating descriptive data to dataset
---

In [8]:
col = ['Velocidad del viento (Km/h)']
wind_speed_l12w_values_df = pd.DataFrame(wind_speed_l12w_array, columns=col)
wind_speed_l12w_values_df_describe = wind_speed_l12w_values_df.describe()

In [9]:
wind_speed_l12w_values_df_describe

Velocidad del viento (Km/h)
count                  4398.000000
mean                      1.297827
std                       2.892122
min                       0.000000
25%                       0.000000
50%                       0.000000
75%                       1.814747
max                      68.571429

In [13]:
# Export this descriptive data to comma separated values and java script object notation
wind_speed_l12w_values_df_describe.to_csv('../../../../data/interim/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                              'Wind-Speed_Describe_May-4_August-03_2018.cvs', sep=',', header=True, index=True)
wind_speed_l12w_values_df_describe.to_json('../../../../data/interim/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n'
                                     'Wind-Speed_Describe_May-4_August-03_2018.json')

# 2. Creating Temperature Training and Testing datasets

We have a **`wind_speed_l12w_values_df`** dataset with 4398 samples rows.

In [14]:
print(wind_speed_l12w_values_df.shape)
wind_speed_l12w_values_df.head()

(4398, 1)


Velocidad del viento (Km/h)
0                          0.0
1                          0.0
2                          0.0
3                          0.0
4                          0.0

We'll divide it into two differents datasets:

- Training dataset
- Testing dataset

This is executed through [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function of this way:

`train_test_split` receive as a data parameter a numpy array, we have to turn the 

**`wind_speed_l12w_values_df`** dataframe to numpy array such as follow:

In [15]:
# numpy_wind_speed_l12w_values_df = wind_speed_l12w_values_df.reset_index().values
numpy_wind_speed_l12w_values_df = wind_speed_l12w_values_df.values

In [16]:
numpy_wind_speed_l12w_values_df

array([[0.        ],
       [0.        ],
       [0.        ],
       ...,
       [0.        ],
       [0.        ],
       [4.21052632]])

We compose the following datasets from **`numpy_wind_speed_l12w_values_df`** array :

- `wind_speed_l12w_values_train`, which is the training matrix
- `wind_speed_l12w_values_test`, the testing matrix

We use the [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function to create the training and testing dataset. Their parameters are:

![alt text](https://cldup.com/hukPWvvLxt-3000x3000.png "klearn.model_selection.train_test_split")


- The first parameter **should be an array**, then we pass the **`numpy_wind_speed_l12w_values_df`** which contain all column features of wind speed.

Of this way, we pass all data (4398 samples rows) to from them `wind_speed_l12w_values_train`

and `wind_speed_l12w_values_test` will be created

- The `test_size=0.5` parameter means a 50% division; which means 
that half of the data goes to the test dataset and the other half 
goes to the training dataset.

A good choice for the size of tests **is usually 0.2 ie 20% or 0.25 or even 30%.** 

In some rare cases we will have 40% but almost never 0.5 or 50%

We choose 20% which means that we will have 20% of 9752 samples or observations for the test data set, 

in this case **4390 * 0.2 = 879.6 samples or records for the test data set**


- The `train_size` parameter is the training dataset size.  **`test_size + train_size = 1 or 100%`**, 
then isn't necessary include it, because wheter we include to `test_size = 0.2`, then the remaining 
data will be to `train_size` this means **0.8 or 80%**

This means that **4390 * 0.8 = 3518.4  training dataset samples rows. **


- random_state is a seed or data source for generating random values for the data sets. 

If this parameter is not passed, the data will be generated in a random way, but in the way as by default numpy works them.

### 2.1  Creating Training and testing wind speed datasets
---

- `wind_speed_l12w_values_train`, which is the training matrix
- `wind_speed_l12w_values_test`, the testing matrix

In [17]:
wind_speed_l12w_values_train, wind_speed_l12w_values_test = train_test_split(numpy_wind_speed_l12w_values_df, test_size = 0.2)

In [19]:
# We have 3518,4 rows to luminosity_luxes_train
print(type(wind_speed_l12w_values_train))
print("The dimensionality of wind speed training dataset is: " +'\n' , wind_speed_l12w_values_train.shape)
print('\n')

# And we have 880 rows to luminosity_luxes_test
print(type(wind_speed_l12w_values_test))
print("The dimensionality of wind speed testing dataset is: " +'\n' , wind_speed_l12w_values_test.shape)

<class 'numpy.ndarray'>
The dimensionality of wind speed training dataset is: 
 (3518, 1)


<class 'numpy.ndarray'>
The dimensionality of wind speed testing dataset is: 
 (880, 1)


In this way, the model, as you progressively learn the correlations in the training set, the better the prediction of the results in the test set.

But if the model learns too much from memory the correlations of the training sets,

that is to say, when one learns from memory and does not understand things, then it

will have problems to predict what is happening on the set of tests, because it is

learned for difficult correlations, if the logic is not well understood and you can

not make good predictions. This is called overfitting or overfitting

The really important thing is to understand that we need to have two different datasets

- Training set with which the ML model learns
- Test set, on which we test whether the ML model correctly learned the correlations


---
##  3. Feature Scaling Temperature training and testing dataset  
---

[This article post](http://benalexkeen.com/feature-scaling-with-scikit-learn/  "Feature Scaling with scikit-learn") it's a great reference to explore the features scaling methods
on scikit learn

- `StandardScaler` assume that data is normally distributed at the level of each characteristic or variable. If the data is not normally distributed, it is not the best alternative to use for scaling.  

- `Min-Max Scaler` it is probably the most famous scaling algorithm and what it does is resize the range to leave it in a dimension of 0 to 1 or -1 and 1 (in case there are negative values in the original dataset of input data)

This scale of maximums and minimums works best for cases where standard scaling may not work properly. If the distribution is not Gaussian or the standard deviation is very small the escalation of maximums and minimums is the best idea.
However, it is sensitive to outliers or outliers, so if there are outliers in the data it is better to consider robust scaling.

- `Robust Scaler` it is similar to the previous one of maximums and minimums, only that it uses interquartile ranges instead of maximums and minimums, which makes it robust for the outliers

- `Normalizer` which scales each value, dividing each value by its magnitude in n dimensional spaces for n number of characteristics.


---

We will use the scaling of maximums and minimums to scale the water table data, because the standard deviation is very small, 

it does not have atypical values and it does not follow a normal distribution (you have to check this)

We apply the maximum and minimum scaling. We provide a rank or base scale that **will be between 0 and 1** using an object,

[MinMaxScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html "sklearn.preprocessing.MinMaxScaler") which transforms each characteristic, (in this 

case it will be the columns feature of `Velocidad del viento (Km/h)`) individually according to a given range.

Product of its applicability, generates these attributes in the dataset, already transformed: 

![alt text](https://cldup.com/lTIv4HXgTk-3000x3000.png "sklearn.preprocessing.MinMaxScaler")

### 3.1 We apply maximium and minimum feature scaling to Wind speed training dataset

In [20]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature training data " + '\n', wind_speed_l12w_values_train)

Remember our temperature training data 
 [[0.        ]
 [0.        ]
 [0.        ]
 ...
 [2.00278164]
 [0.        ]
 [2.01398601]]


With the [fit](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.fit  "MinMaxScaler.fit")
we compute the maximum and minimum value of  `wind_speed_l12w_values_train` dataset to be used in the subsequent scaling 

We assing these values to `minmax_scale_training` variable.

In [21]:
minmax_scale_training = scaler.fit(wind_speed_l12w_values_train.astype(float))
# print(minmax_scale_training.data_max_)
# http://terrapinssky.blogspot.com/2017/10/pythonresolved-dataconversionwarning.html

Then, we apply the [transform](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.transform "MinMaxScaler.transform") method to transform these data to maximum and mínimum scale value. 

Here, with this process, the `wind_speed_l12w_values_train` data are scaled between **0 to 1**  selected range

In [22]:
# transform precipitations_train data to maximum and mínimum scale value. 
wind_speed_l12w_minmax_training = minmax_scale_training.transform(wind_speed_l12w_values_train)

In [23]:
print ("And, these are our scaled data: " + '\n')
wind_speed_l12w_minmax_training

And, these are our scaled data: 



array([[0.        ],
       [0.        ],
       [0.        ],
       ...,
       [0.02920723],
       [0.        ],
       [0.02937063]])

In [24]:
print('Wind speed Training dataset. Minimum value after MaxMinScaler:\nWind speed={:.1f}'
      .format(wind_speed_l12w_minmax_training[:,0].min()))

print('Wind speed Training dataset. Maximum value after MaxMinScaler:\nWind speed={:.1f}'
      .format(wind_speed_l12w_minmax_training[:,0].max()))

Wind speed Training dataset. Minimum value after MaxMinScaler:
Wind speed=0.0
Wind speed Training dataset. Maximum value after MaxMinScaler:
Wind speed=1.0


Then, our  MinMaxScaler normalized training dataset is `wind_speed_l12w_minmax_training` numpy array

- We export this array to comma separated values 

In [25]:
wind_speed_l12w_train_df = pd.DataFrame(wind_speed_l12w_minmax_training, columns=col)

In [28]:
print(wind_speed_l12w_train_df.shape)
wind_speed_l12w_train_df.head()

(3518, 1)


Velocidad del viento (Km/h)
0                     0.000000
1                     0.000000
2                     0.000000
3                     0.000000
4                     0.075812

In this way we have the dataset **`wind_speed_l12w_train_df`** standardized and training, and export it to a .csv file

In [31]:
wind_speed_l12w_train_df.to_csv('../../../../data/processed/WindSpeed/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                                 'Wind-Speed_Normalized_TRAINING_May-4_August-03_2018.csv', sep=',', header=True, index=False)

### 3.2 We apply maximium and minimum feature scaling to Temperature testing dataset


In [32]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature testing data " + '\n', wind_speed_l12w_values_test)

Remember our temperature testing data 
 [[ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 1.23287671]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 2.68656716]
 [ 0.        ]
 [ 2.74285714]
 [ 0.        ]
 [ 7.4611399 ]
 [ 0.        ]
 [ 0.        ]
 [ 1.23924269]
 [ 4.1025641 ]
 [ 0.        ]
 [ 3.82978723]
 [ 1.47088866]
 [ 2.05420827]
 [ 0.        ]
 [ 0.        ]
 [ 1.16978067]
 [ 0.        ]
 [ 1.31028207]
 [ 0.        ]
 [ 3.71134021]
 [ 1.3832853 ]
 [ 0.        ]
 [ 3.39622642]
 [ 1.06116433]
 [ 1.81818182]
 [ 0.        ]
 [ 0.        ]
 [ 1.94070081]
 [ 1.61797753]
 [ 1.20502092]
 [ 1.1136891 ]
 [ 0.        ]
 [ 0.        ]
 [ 8.13559322]
 [ 0.        ]
 [ 3.11015119]
 [ 0.        ]
 [ 2.25705329]
 [ 1.19205298]
 [ 6.31578947]
 [ 1.81588903]
 [ 2.13017752]
 [ 5.58139535]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 1.50943396]
 [ 0.        ]
 [ 1.82278481]
 [ 0.        ]


In [33]:
minmax_scale_test = scaler.fit(wind_speed_l12w_values_test.astype(float))
# transform luminosity_luxes_test data to maximum and mínimum scale value. 
wind_speed_l12w_minmax_test = minmax_scale_test.transform(wind_speed_l12w_values_test)

In [34]:
print ("And, these are our testing scaled data: " + '\n')
wind_speed_l12w_minmax_test

And, these are our testing scaled data: 



array([[0.        ],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.02482877],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.05410448],
       [0.        ],
       [0.0552381 ],
       [0.        ],
       [0.15025907],
       [0.        ],
       [0.        ],
       [0.02495697],
       [0.08262108],
       [0.        ],
       [0.07712766],
       [0.02962206],
       [0.04136947],
       [0.        ],
       [0.        ],
       [0.02355808],
       [0.        ],
       [0.02638763],
       [0.        ],
       [0.07474227],
       [0.02785783],
       [0.        ],
       [0.06839623],
       [0.02137067],
       [0.03661616],
       [0.        ],
       [0.        ],
       [0.03908356],
       [0.03258427],
       [0.02426778],
       [0.02242846],
       [0.        ],
       [0.        ],
       [0.16384181],
       [0.        ],
       [0.06263499],
       [0.        ],
       [0.04545455],
       [0.024

In [35]:
print('Wind Speed Testing dataset. Minimum value after MaxMinScaler:\nWind Speed={:.1f}'
      .format(wind_speed_l12w_minmax_test[:,0].min()))

print('Wind Speed Testing dataset. Maximum value after MaxMinScaler:\nWind Speed={:.1f}'
      .format(wind_speed_l12w_minmax_test[:,0].max()))

Wind Speed Testing dataset. Minimum value after MaxMinScaler:
Wind Speed=0.0
Wind Speed Testing dataset. Maximum value after MaxMinScaler:
Wind Speed=1.0


Then, our  MinMaxScaler normalized training dataset is `wind_speed_l12w_minmax_test` numpy array

- We export this array to comma separated values

In [36]:
wind_speed_l12w_test_df = pd.DataFrame(wind_speed_l12w_minmax_test, columns=col)

In [37]:
print(wind_speed_l12w_test_df.shape)
wind_speed_l12w_test_df

(880, 1)


Velocidad del viento (Km/h)
0                       0.000000
1                       0.000000
2                       0.000000
3                       0.000000
4                       0.024829
5                       0.000000
6                       0.000000
7                       0.000000
8                       0.000000
9                       0.054104
10                      0.000000
11                      0.055238
12                      0.000000
13                      0.150259
14                      0.000000
15                      0.000000
16                      0.024957
17                      0.082621
18                      0.000000
19                      0.077128
20                      0.029622
21                      0.041369
22                      0.000000
23                      0.000000
24                      0.023558
25                      0.000000
26                      0.026388
27                      0.000000
28                      0.074742
29                      0.027858
..                           ...
850                     0.000000
851                     0.133641
852                     0.077540
853                     0.045031
854                     0.036616
855                     0.109848
856                     0.020337
857                     0.000000
858                     0.000000
859                     0.061181
860                     0.050967
861                     0.050699
862                     0.038462
863                     0.048414
864                     0.053803
865                     0.000000
866                     0.024106
867                     0.000000
868                     0.046850
869                     0.000000
870                     0.082857
871                     0.000000
872                     0.000000
873                     0.000000
874                     0.024227
875                     0.036802
876                     0.000000
877                     0.000000
878                     0.039835
879                     0.000000

[880 rows x 1 columns]

In [ ]:
temperature_l12w_test_df.to_csv('../../../../data/processed/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                                 'Temperature_Normalized_TESTING_May-4_August-03_2018.csv', sep=',', header=True, index=False)